### PR TITLE
Modularize distance and area formatting

### PR DESCRIPTION
**Pull request by m-kuhn**
All testing done so far is to validate, that it compiles :)

### CODE REVIEW

**Comment by nyalldawson** [comment]
Looks good! Do you think scaled distance/... would belong better in QgsUnitTypes? (I do). 

I think there's a lot more code which needs updating when a new distance unit is added too...

File: src/core/qgsunittypes.h [hunk 1]
```diff
@@ -49,6 +49,8 @@ class CORE_EXPORT QgsUnitTypes
       DistanceYards, //!< Imperial yards
       DistanceMiles, //!< Terrestrial miles
       DistanceDegrees, //!< Degrees, for planar geographic CRS distance measurements
+      DistanceCentimeters, //! Centimeters
```
Comment (nyalldawson) [inline src/core/qgsunittypes.h]: Missing < on the docstrings

**Comment by m-kuhn** [comment]
> Looks good! Do you think scaled distance/... would belong better in QgsUnitTypes? (I do).

Why not... I think it belongs next to `formatXyz`, `convertXyzMeasurement`, so they should probably be moved too.

**Comment by nyalldawson** [comment]
Agreed. I just don't think most people would look in QgsDistanceArea for these functions.

Off topic: the name of QgsDistanceArea really bugs me. Shouldn't it be QgsDistanceAreaCalculator or QgsDistanceAreaMeasurer instead?

**Comment by m-kuhn** [comment]
True, `QgsDistanceArea` implies storage of distances and areas. What I also don't like very much is that it lists the two terms "distance" and "area" and will need to be renamed when we get "volumes" 📦 . So maybe `QgsMeasurementCalculator` would be an alternative?

**Comment by nyalldawson** [comment]
Can't help but think this PR has got a little side tracked ;)

**Comment by m-kuhn** [comment]
@nyalldawson should be back on track now. Anything you'd like to add?

**Comment by nyalldawson** [comment]
All looks good to me - there's just some tests in test_qgsunittypes.py which need to be updated to add the new units. Just scroll through the existing tests and add cm/mm following the existing test setup.

File: src/core/qgsunittypes.cpp [hunk 2]
```diff
@@ -864,6 +979,12 @@ QgsUnitTypes::AreaUnit QgsUnitTypes::distanceToAreaUnit( DistanceUnit distanceUn
     case DistanceKilometers:
       return AreaSquareKilometers;
 
+    case DistanceCentimeters:
```
Comment (nyalldawson) [inline src/core/qgsunittypes.cpp]: This isn't right... I think you'll need to unfortunately add AreaSquareMillimeters/AreaSquareCentimeters to correctly handle this.

**Comment by m-kuhn** [comment]
@nyalldawson all concerns should be addressed now.
Merge?

File: src/core/qgis.h [hunk 3]
```diff
@@ -233,6 +233,17 @@ inline double qgsRound( double x )
...
 }
 
+/**
+ * Returns a double \a number, rounded (as close as possible) to the specified number of \a places.
+ *
+ * @note Added in QGIS 3.0
+ */
+inline double qgsRound( double number, double places )
+{
+  int scaleFactor = pow( 10, places );
```
Comment (agiudiceandrea) [inline src/core/qgis.h]: @m-kuhn, I think there is a flaw in qgsRound() that leads to incorrect rounding and/or overflow in some circumstances.
Could you please have a look at http://osgeo-org.1560.x6.nabble.com/QGIS-Developer-On-precision-in-qgsRound-3-3-0-master-td5377378.html and https://github.com/qgis/QGIS/pull/7836#issuecomment-419718554

**Pull request merged**